### PR TITLE
Added paged documents links handling to APIProvider

### LINKED ISSUE #187: Add documentation about pagination
It might not be clear from the get-go how to use pagination support. Adding this to the readme would help implementors!

### PR DESCRIPTION
**Pull request by andyj-at-aspin**
Here's a new API for handling paged sets of results from AppStore Connect.

The additions do not rely on chopping up bits of URL to find specific URL Query parameters, rather they access the entire URL passed to the client as the next page link by AppStore Connect, adding in the necessary JWT authentication headers and decoding the subsequent response in the same way a direct call to  APIProvider.request(_) would  decode the response from the endpoint. As such, these functions do not need to care what the endpoint is to use its generic specialisation to apply the correct decode() and return the correct structure.

```swift
    let request = APIEndpoint.v1.apps
        .get(parameters: .init(
            sort: [.bundleID],
            fieldsApps: [.appInfos, .name, .bundleID],
            limit: 2
        ))
    
    // Demonstration of AsyncSequence result of APIProvider.paged(_)
    var allApps: [App] = []
    for try await pagedResult in provider.paged(request) {
        allApps.append(contentsOf: pagedResult.data)
    }

    // Demonstration of APIProvider.hasPagedDocumentLinks(_) and APIProvider.request(_: pageAfter:)
    let firstPageResult = try await provider.request(request)
    let firstPageApps = firstPageResult.data
    print(firstPageApps)
    if provider.hasPagedDocumentLinks(firstPageResult) {        
        if let nextPage = try await provider.request(request, pageAfter: firstPageResult) {
            let secondPageApps = nextPage.data
            print(secondPageApps)
        }
    }

```

APIProvider has new public functions:

    // Does this object have a property that is a PagedDocumentLinks structure?
    public func hasPagedDocumentLinks<T>(_ object: T) -> Bool 

    // If the argument pageAfter has a PagedDocumentLinks.next property, get the next page of results, 
    // decoding the response as if it had been a direct call to the requested endpoint (async version)
    public func request<T>(_ endpoint: Request<T>, pageAfter: T) async throws -> T? 

    // If the argument pageAfter has a PagedDocumentLinks.next property, get the next page of results, 
    // decoding the response as if it had been a direct call to the requested endpoint (completion 
    // handler version)
    public func request<T: Decodable>(_ request: Request<T>, pageAfter currentPage: T, completion: @escaping RequestCompletionHandler<T?>) 

    // Return all pages of responses to the endpoint as an AsyncSequence
    public func paged<T>(_ endpoint: Request<T>) -> PagedRequest<T>  

Fixes #187

### CODE REVIEW

File: Sources/APIProvider.swift [hunk 1]
```diff
@@ -274,3 +274,73 @@ private extension APIProvider {
     }
 
 }
+
+public struct PagedRequest<T>: AsyncSequence, AsyncIteratorProtocol where T: Decodable {
```
Comment (AvdLee) [inline Sources/APIProvider.swift:278]: I love how you adopted AsyncSequence for this! The implementation side for users look amazing due to this 💪

Comment (AvdLee) [reply Sources/APIProvider.swift:278]: Would you mind moving this into a separate file?

File: Sources/APIProvider.swift [hunk 2]
```diff
@@ -274,3 +274,73 @@ private extension APIProvider {
...
+        let mirror = Mirror(reflecting: object)
+        return mirror.children.first(where: { $0.value is PagedDocumentLinks }).flatMap({ $0.value as? PagedDocumentLinks })
+    }
+
+    public func request<T: Decodable>(_ request: Request<T>, pageAfter currentPage: T, completion: @escaping RequestCompletionHandler<T?>) {
+                        
+        guard
+            let nextPage = pagedDocumentLinks(currentPage)?.next,
+            let url = URL(string: nextPage)
+        else { completion(.success(nil)); return }
```
Comment (AvdLee) [inline Sources/APIProvider.swift:338]: Nit: I prefer to keep things on separate lines

```suggestion
        else { 
            completion(.success(nil))
            return 
        }
```

File: Sources/APIProvider.swift [hunk 3]
```diff
@@ -274,3 +274,73 @@ private extension APIProvider {
...
+        }
+    }
+
+    public func hasPagedDocumentLinks<T>(_ object: T) -> Bool {
+        pagedDocumentLinks(object) != nil
+    }
+    
+    private func pagedDocumentLinks<T>(_ object: T) -> PagedDocumentLinks? {
+        let mirror = Mirror(reflecting: object)
+        return mirror.children.first(where: { $0.value is PagedDocumentLinks }).flatMap({ $0.value as? PagedDocumentLinks })
```
Comment (AvdLee) [inline Sources/APIProvider.swift:330]: Super smart, I love how this makes it work with just any endpoint!

**Review by AvdLee — CHANGES_REQUESTED** [review]
Thanks a lot for this, I couldn't have done it in a much nicer way, well done!! I just have a few points of feedback that should be easy to solve

**Comment by AvdLee** [comment]
Also, @andyj-at-aspin, it would be great to include your PR description example in a section to the readme!

**Comment by andyj-at-aspin** [comment]
Ok. Done all that Antoine. 

I have changed the function call associated with evaluating whether an object had PagedDocumentLinks properties so one that required the request to be passed in as well - so enforcing the test was only run on expected endpoint response types. 

I have added the examples to the readme and have proved to myself that the example still works with the change to the function call.

I've added some documentation to the new public function declarations, but if they're no good I apologise. 

I've never made project changes following a pull request before and I don't know of this is all I need to do or if I have to do some further action for you to see the new commit. I work in a team of 2 - both with 25+ years experience in the same company - and we are responsible for merging our own work into the relevant development/release branches.

**Review by AvdLee — APPROVED** [review]
You've actually done great! I'm super happy with this contribution, thanks a lot! 🎉

**Pull request merged**